### PR TITLE
Decypher batches wihtin enclave

### LINKED ISSUE #454: Improve getDecryptionShares call
Currently this call is decyphering each ciphertext within the enclave at a time. Private key share is being decyphered for each ciphertext within the batch.

We should decypher the private key once, and use it to decypher the batch.

### PR DESCRIPTION
**Pull request by PropzSaladaz**
### Description

- We now concatenate each received ciphertext into a single string
- We parse ciphertexts from this concatenated string in a BATCH basis
    - get a batched size of cyphertexts & send it to enclave at a time
    - This BATCH is fixed-size (we pre-allocate the memory within enclave)
    - If total ciphertexts > BATCH, then decypher the remainder again using enclave
    - This means we will repeat the decyphering of the private key for each batch processed within the enclave
- Added new condition to check on decryptions array size. Could be a source of attack (if array was larger than int size, for loop would run forever, as it would wrap around int size, and never reach array size)
- The response now includes:
  - `publicDecryptionValues` : array with all the decryptions shares, by the same order they were requested. If one of the shares to decrypt is invalid, then the corresponding response for that share will be 256 zeroes, and the index of the request will be present in the `failedRequests` field.
  - `failedRequests`: A map of `index` -> `status error`. Includes only the indices that failed, with the respective failure error.
  - If no failed requests, then the field `failedRequests` is ommitted
    
### Tests
- Tested locally making batched requests, & validating the results with libBLS
- Added test edge cases
    - Num. cyphertexts < BATCH size
    - Num. cyphertexts == BATCH size
    - Num. cyphertexts > BATCH size (but not multiple of it)
    - Num. cyphertexts > BATCH size (multiple of it)
 - Updated the tests regarding `getDecryptionShares` to use the new interface (4 64-character hexadecimals encoding the point coordinates)
 
 ### Performance Tests
 The results below were the result of repeating each test 30 times & computing the mean.
 | **Num. shares to decrypt sent in a single request** | **Old (tx / second)** | **New (tx / second)** | **% increase** |
| --- | --- | --- | --- |
| 20 | 501 | 502 | 0.2 |
| 50 | 531 | 535 | 0.7 |
| 100 | 546 | 562 | 2.9 |
| 200 | 548 | 566 | 3.3 |
| 400 | 551 | 570 | 3.4 |
| 800 | 557 | 570 | 2.3 |

Improved on average ~2-3%

A note on further improvements - With new implementation, the SGX side needs to concatenate all decryption shares to pass it to enclave as `char *`. If request already comes as a single concatenated string, we can offload this work to the client of SGX and may improve throughput.

Fixes #454

### CODE REVIEW

File: secure_enclave/EnclaveConstants.h [hunk 1]
```diff
@@ -26,6 +26,11 @@
 
 #define BUF_LEN 1024
 
+#define BATCH_SIZE 100
+#define CIPHERTEXT_LEN 256
```
Comment (olehnikolaiev) [inline secure_enclave/EnclaveConstants.h]: lets make the names more descriptive. `BATCH_SIZE` may refer to many things and it is unclear what is it's usage

File: testw.cpp [hunk 2]
```diff
@@ -1364,44 +1359,28 @@ TEST_CASE_METHOD(TestFixture,
...
+        client->getDecryptionShares(name, publicDecryptionValues);
 
-  REQUIRE(share2 == key * decryption_value2);
+    for (int i = 0; i < num_requests; i++) {
+      auto decryption_share = decryptionShares[i].asString();
+      libff::alt_bn128_G2 share = convertStringToG2(decryption_share);
+      REQUIRE(share == key * decryption_values[i]);
+    }
+  }
 }
```
Comment (olehnikolaiev) [inline testw.cpp:1386]: are there any test that verify that `getDecryptionShares` method does not crash on incorrect input? if no lets add it

File: SGXWalletServer.cpp [hunk 3]
```diff
@@ -1112,22 +1112,32 @@ Json::Value SGXWalletServer::getDecryptionSharesImpl(
...
+      throw SGXException(INVALID_DECRYPTION_VALUE_FORMAT,
+                         string(__FUNCTION__) +
+                             ":Public decryption values array is too large");
+    }
+
+    shared_ptr<string> encryptedKeyHex_ptr = readFromDb(blsKeyName);
+
+    // validate & concatenate ciphertexts
+    std::string concatenatedCiphertexts;
+    concatenatedCiphertexts.reserve(MAX_BATCH_SIZE * 256 + 1);
```
Comment (olehnikolaiev) [inline SGXWalletServer.cpp]: lets use `CIPHERTEXT_LEN` here

File: TECrypto.cpp [hunk 4]
```diff
@@ -39,10 +39,25 @@
...
   size_t sz = 0;
 
+  // calculate number of batches needed
+  size_t numBatchesRemaining = decryptionValueBatches.size() / BATCH_SIZE_BYTES;
+  size_t lastBatchRemainderBytes = 0;
+  bool firstBatchIsFull = decryptionValueBatches.size() >= BATCH_SIZE_BYTES;
+  // There is at least +1 cyphertext that does not fit in the batch count we
+  // currently have - increase batch count
+  if (decryptionValueBatches.size() % BATCH_SIZE_BYTES != 0) {
+    numBatchesRemaining++;
```
Comment (olehnikolaiev) [inline TECrypto.cpp]: isnt it the same as
```
size_t numBatchesRemaining = ( decryptionValueBatches.size() + BATCH_SIZE_BYTES - 1 )/ BATCH_SIZE_BYTES;

size_t lastBatchRemainderBytes = decryptionValueBatches.size() % BATCH_SIZE_BYTES;
```

File: TECrypto.cpp [hunk 5]
```diff
@@ -52,22 +67,49 @@ vector<string> calculateDecryptionShare(const string &encryptedKeyShare,
...
+    // split the decrypted shares into individual shares
+    for (size_t i = 0; (i < MAX_BATCH_BUF_SIZE) && (i < decr_shares.length());
+         i += CIPHERTEXT_LEN) {
+      decryptedBatches.push_back(decr_shares.substr(i, CIPHERTEXT_LEN));
+    }
+
+    --numBatchesRemaining;
+    // advance batch
+    currentBatch += BATCH_SIZE_BYTES;
+    currentBatchOffset += BATCH_SIZE_BYTES;
```
Comment (olehnikolaiev) [inline TECrypto.cpp]: is`currentBatchOffset` used anywhere else?

File: TECrypto.cpp [hunk 6]
```diff
@@ -52,22 +67,49 @@ vector<string> calculateDecryptionShare(const string &encryptedKeyShare,
...
-      std::make_shared<std::string>(decryptionShare), ":");
+    // split the decrypted shares into individual shares
+    for (size_t i = 0; (i < MAX_BATCH_BUF_SIZE) && (i < decr_shares.length());
+         i += CIPHERTEXT_LEN) {
+      decryptedBatches.push_back(decr_shares.substr(i, CIPHERTEXT_LEN));
+    }
+
+    --numBatchesRemaining;
+    // advance batch
+    currentBatch += BATCH_SIZE_BYTES;
```
Comment (olehnikolaiev) [inline TECrypto.cpp]: lets use pointer arithmetics carefully. need to check here that pointer does not go out of bounds

File: secure_enclave/TEUtils.cpp [hunk 7]
```diff
@@ -50,10 +50,15 @@
...
 
+/**
+ * Converts a field to hexadecimal format.
+ * By default, output a 64-character string (32 bytes - 2 hexadecimal characters
+ * per byte)
+ */
 template <class T>
-std::string fieldElementToString(const T &field_elem, int base = 10) {
-
+std::string fieldElementToString(const T &field_elem, int numBytes = 32) {
```
Comment (olehnikolaiev) [inline secure_enclave/TEUtils.cpp]: lets rename to `fieldElementToHex` then

File: secure_enclave/TEUtils.cpp [hunk 8]
```diff
@@ -65,152 +70,147 @@ std::string fieldElementToString(const T &field_elem, int base = 10) {
...
+  return output;
 }
 
-std::vector<libff::alt_bn128_Fq> SplitStringToFq(const char *coords,
-                                                 const char symbol) {
-  std::vector<libff::alt_bn128_Fq> result;
-  std::string str(coords);
-  std::string delim;
+libff::alt_bn128_G2 stringToG2(char *str) {
+  // if ( str.size() != 256 ) {
```
Comment (olehnikolaiev) [inline secure_enclave/TEUtils.cpp]: dont we need this check?

File: secure_enclave/TEUtils.cpp [hunk 9]
```diff
@@ -65,152 +70,147 @@ std::string fieldElementToString(const T &field_elem, int base = 10) {
...
-        libff::alt_bn128_Fq coeff(token.c_str());
-        result.push_back(coeff);
-      }
-      prev = pos + delim.length();
-    } while (pos < str.length() && prev < str.length());
+  return ret;
+}
 
-    return result;
+EXTERNC int keyHexToDecimal(char *skey_hex) {
```
Comment (olehnikolaiev) [inline secure_enclave/TEUtils.cpp]: lets not override pointers and add a second parameter for output data

File: TECrypto.cpp (hunk 6 above)
Comment (PropzSaladaz) [reply TECrypto.cpp]: I think the only time it can logically go out of bounds is at the end of the last iteration. Since it is the last iteration, we never access it again, but it is still incremented. Should I still add a condition right after the `--numBatchesRemaining` to avoid incrementing the pointer in that case?

For all other iterations, the pointer is always valid. We compute the number of batches initially, and increase this pointer on a batch size basis. If the last batch is < BATCH_SIZE, we set the lastBatchRemainderBytes & effectively only read that amount of bytes from the pointer, respecting the string's initial size.

**Review by olehnikolaiev — DISMISSED** [review]

File: DKGCrypto.cpp [hunk 10]
```diff
@@ -37,16 +37,24 @@
...
 
   SAFE_CHAR_BUF(arr, mpz_sizeinbase(t, base) + 2);
 
   mpz_get_str(arr, base, t);
 
+  result = arr;
+
+  if (base == 16) {
+    // 64-characters long - fill 0's if needed
+    int n_zeroes = 64 - result.length();
```
Comment (kladkogex) [inline DKGCrypto.cpp]: Hey Sidnei

It is better here to add a check that result.length() is less than 64, otherwise there will be an overflow.

In general functions should check preconditions on inputs, so if there is an error the functions returns and error/throws an exception instead of crashing

File: secure_enclave/TEUtils.cpp [hunk 11]
```diff
@@ -65,152 +70,150 @@ std::string fieldElementToString(const T &field_elem, int base = 10) {
...
     LOG_ERROR(e.what());
-    goto clean;
   } catch (...) {
     LOG_ERROR("Unknown throwable");
-    goto clean;
   }
 
-clean:
   mpz_clear(t);
   return ret;
```
Comment (kladkogex) [inline secure_enclave/TEUtils.cpp]: Here if an error occurs, the exception is caught and potentially invalid value ret is returned.

How does the caller of the function know that an error occured, so it can be handled?

File: secure_enclave/TEUtils.cpp [hunk 12]
```diff
@@ -65,152 +70,150 @@ std::string fieldElementToString(const T &field_elem, int base = 10) {
...
+    char *tmp = mpz_get_str(arr, 10, dec);
+
+    output = tmp;
   } catch (std::exception &e) {
     LOG_ERROR(e.what());
-    return result;
   } catch (...) {
-    LOG_ERROR("Unknown throwable");
-    return result;
+    LOG_ERROR("Exception in convert hex to dec");
```
Comment (kladkogex) [inline secure_enclave/TEUtils.cpp:151]: Same comment as above - how does the caller of the function find out that an error occured?

File: secure_enclave/TEUtils.cpp [hunk 13]
```diff
@@ -65,152 +70,150 @@ std::string fieldElementToString(const T &field_elem, int base = 10) {
...
-        libff::alt_bn128_Fq coeff(token.c_str());
-        result.push_back(coeff);
-      }
-      prev = pos + delim.length();
-    } while (pos < str.length() && prev < str.length());
+    if (mpz_set_str(skey, skey_hex, 16) == -1) {
+      mpz_clear(skey);
+      LOG_ERROR("Could not convert hexadecimal into number");
+      return 1;
+    }
```
Comment (kladkogex) [inline secure_enclave/TEUtils.cpp:188]: Why is 1 returned here? In general the function should either throw an exception or return an error code as a separate variable

File: CryptoTools.cpp [hunk 14]
```diff
@@ -63,11 +63,12 @@ bool hex2carray(const char *_hex, uint64_t *_bin_len, uint8_t *_bin,
   CHECK_STATE(_bin);
   CHECK_STATE(_bin_len)
 
-  uint64_t len = strnlen(_hex, 2 * _max_length + 1);
+  size_t max_length_2 = 2 * _max_length;
+  uint64_t len = strnlen(_hex, max_length_2 + 1);
 
-  CHECK_STATE(len != 2 * _max_length + 1);
+  CHECK_STATE(len != max_length_2 + 1);
```
Comment (kladkogex) [inline CryptoTools.cpp:69]: This statement is not needed since the next statement checks that 

len <= max_length_2

File: secure_enclave/TEUtils.cpp (hunk 11 above)
Comment (kladkogex) [reply secure_enclave/TEUtils.cpp]: In general, in case of an error a function should

1) clean resources so there is no memory leak
2) either return an error as a separate variable or throw an exception


Please check the pull request for this )

File: DKGCrypto.cpp (hunk 10 above)
Comment (PropzSaladaz) [reply DKGCrypto.cpp]: In this case, since libff types are 32 bytes, it should never exceed 64 characters. But still, I'll add it for safety.

File: CryptoTools.cpp (hunk 14 above)
Comment (PropzSaladaz) [reply CryptoTools.cpp:69]: Nice catch. I'll leave the 1st one and get rid of the second, since the first one is stronger / stricter

**Review by kladkogex — DISMISSED** [review]

**Review by olehnikolaiev — DISMISSED** [review]

**Review by kladkogex — DISMISSED** [review]

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]
Please, bump SGX version to 1.10.0

**Review by olehnikolaiev — APPROVED** [review]

**Review by DmytroNazarenko — APPROVED** [review]
